### PR TITLE
Add StartupHooks to Plugins

### PR DESCRIPTION
**Pull request by hacdias**
StartupHook enables the developer to create plugins that are executed in the very beginning of the Caddy process, even before it has started running. It allows, among many other things, to use flags and take advantage of the fact that Caddy hasn't started running yet to configure and enable a variety of features. I think this hook will open a whole new range of plugins.

I've already talked with @mholt about this because I was having a problem while I wanted to develop a plugin and he suggested this.

### CODE REVIEW

File: plugins.go [hunk 1]
```diff
@@ -176,6 +200,10 @@ type Plugin struct {
 	// Action is the plugin's setup function, if associated
 	// with a directive in the Caddyfile.
 	Action SetupFunc
+
+	// StartupHook is the plugin's function that is executed
+	// imediatelly after the flag parsing.
```
Comment (abiosoft) [inline plugins.go]: typo, should be `immediately`

File: plugins.go [hunk 2]
```diff
@@ -69,6 +69,30 @@ func DescribePlugins() string {
 	return str
 }
 
+// StartupHooks executes the startup hooks defined when the
+// plugins where registred and returns the first error
```
Comment (abiosoft) [inline plugins.go]: typo, should be `were registered`

**Comment by abiosoft** [comment]
Well done @hacdias 👏

**Comment by hacdias** [comment]
Thanks @abiosoft! Typos corrected :smile:

File: plugins.go [hunk 3]
```diff
@@ -69,6 +69,30 @@ func DescribePlugins() string {
 	return str
 }
 
+// StartupHooks executes the startup hooks defined when the
+// plugins were registred and returns the first error
```
Comment (abiosoft) [inline plugins.go]: There is still one typo, regist**e**red.

**Comment by hacdias** [comment]
@abiosoft thanks 😄

**Comment by wendigo** [comment]
LGTM

File: caddy/caddymain/run.go [hunk 4]
```diff
@@ -97,6 +97,12 @@ func Run() {
 		mustLogFatalf(err.Error())
 	}
 
+	// Execute Startup Hooks
```
Comment (mholt) [inline caddy/caddymain/run.go]: A more useful comment might be, "Execute plugins that are registered to run as the process starts"

File: caddy/caddymain/run.go [hunk 5]
```diff
@@ -97,6 +97,12 @@ func Run() {
...
+	err = caddy.StartupHooks(serverType)
+	if err != nil {
+		mustLogFatalf(err.Error())
+	}
```
Comment (mholt) [inline caddy/caddymain/run.go]: Change to `"%v", err` just to be safe (I know, I know, we don't do that anywhere else in this file, I just thought of it.)

**Comment by hacdias** [comment]
Sweet 😄

**Pull request merged**